README
python-yourls is a simple python based API client for the YOURLS URL shortener
(http://yourls.org/).

For more information on using the client, documentation has been written for
sphinx (http://sphinx.pocoo.org/) and it can be built with:

  $ python setup.py build_sphinx

Current documentation is available at http://tflink.github.com/python-yourls


docs/changelog.rst
.. _changelog:

Changelog
=========================================

Version 0.3.0 (2012-09-13)
---------------------------
 * Command Line interface
 * Added commands to support Yourls 1.5.1
 * Refactored tests using unittest
 * Minor bugfixes

Version 0.2.0 (2011-11-21)
---------------------------
 * Greatly improved error handling
 * Changed behavior and name of errors raised
   * YourlsKeywordError is now YourlsOperationError and is thrown in more places

Version 0.1.2 (2011-11-14)
---------------------------
 * Fixed Tests
 * Added support for the 'title' parameter (thanks to Davmuz)
 * Fixed issue where using an empty string as the api url wouldn't raise an
   exception at creation time

Version 0.1.1 (2011-08-25)
---------------------------
 * Initial Release

.. toctree::
   :maxdepth: 2

..    main

Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`



tests/test_yourlsClient.py
# test_yourlsClient.py
#  - tests for the python yourls client
#
# Copyright 2011, Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Author:
#       Tim Flink <tflink@redhat.com> - original test unit
#       Jordi Llonch <llonchj@gmail.com> - test refactoring using unittest

import os
import json
import string
import random
import httplib
import urlparse
import unittest

import yourls.client
from yourls import YourlsError, YourlsOperationError

test_baseurl = os.environ.get('YOURLS_TEST_BASEURL', 
    'http://localhost/yourls/')
test_apiurl = os.environ.get('YOURLS_TEST_APIURL', 
    urlparse.urljoin(test_baseurl, 'yourls-api.php'))
    
test_user = os.environ.get('YOURLS_TEST_USER', 'username')
test_pass = os.environ.get('YOURLS_TEST_PASSWORD', 'password')
test_token = os.environ.get('YOURLS_TEST_TOKEN', None)

def keyword_generator(size=15, chars=string.ascii_lowercase + string.digits):
    return ''.join(random.choice(chars) for x in range(size))

def setUp():
    print "Using test server..."
    print "API Url: %s" % test_apiurl
    print "Username: %s" % test_user
    print "Password: %s" % ("*" * len(test_pass))
    print
    if test_token is None:
        msg = " Define environ 'YOURLS_TEST_TOKEN' to test PasswordLess API access "
        print "*" * len(msg)
        print msg
        print "*" * len(msg)
        print

class YourlsTest(unittest.TestCase):
    url = "http://www.google.com/"
    
    def shorturl_and_expand(self, client, url=None):
        if url is None:
            url = self.url

        keyword = keyword_generator()

        result = client.shorturl(url, keyword=keyword)
        return client.expand(result)
        
    def browse(self, url):
        parsed_url = urlparse.urlparse(url)
        conn = httplib.HTTPConnection(parsed_url.hostname)
        conn.request("GET", parsed_url.path)
        return conn.getresponse()

    def test_missing_credentials(self):
        self.assertRaises(YourlsError, 
            yourls.client.YourlsClient, (test_apiurl,))

    def test_missing_username(self):
        self.assertRaises(YourlsError, 
            yourls.client.YourlsClient, (test_apiurl, None, test_pass))

    def test_missing_password(self):
        self.assertRaises(YourlsError, 
            yourls.client.YourlsClient, (test_apiurl, test_user, None))

    def test_invalid_yourls_server(self):
        self.assertRaises(YourlsError, 
            yourls.client.YourlsClient, (None,))

    def test_invalid_token(self):
        self.assertRaises(YourlsError, 
            yourls.client.YourlsClient, (test_apiurl,), 
            dict(token=str(test_token) + "xxx"))

    def test_invalid_user(self):
        client = yourls.client.YourlsClient(test_apiurl, 
            username=test_user + "xxx", password=test_pass)
        self.assertRaises(YourlsOperationError, client.shorturl, (self.url,))

    def test_invalid_password(self):
        client = yourls.client.YourlsClient(test_apiurl, 
            username=test_user, password=test_pass + "xxx")
        self.assertRaises(YourlsOperationError, client.shorturl, (self.url,))

    def test_non_existing_keyword(self):
        client = yourls.client.YourlsClient(test_apiurl, 
            username=test_user, password=test_pass)
        for i in range(1, 10):
            keyword = keyword_generator(30)
            self.assertRaises(YourlsOperationError, 
                client.expand, (keyword,))

    def test_token_shorturl_and_expand(self):
        assert test_token

        client = yourls.client.YourlsClient(test_apiurl, token=test_token)
        self.assertEqual(self.url, self.shorturl_and_expand(client))

    def test_login_shorturl_and_expand(self):
        client = yourls.client.YourlsClient(test_apiurl, 
            username=test_user, password=test_pass)
        self.assertEqual(self.url, self.shorturl_and_expand(client))

    def test_stats(self):
        client = yourls.client.YourlsClient(test_apiurl, 
            username=test_user, password=test_pass)
        
        keyword = keyword_generator()
        url = "http://www.gmail.com/"
        title = "i am the title"

        shorturl = client.shorturl(url, keyword=keyword, title=title)
        
        result = client.url_stats(shorturl)

        self.assertTrue(isinstance(result, dict))

        #
        self.assertEqual(result.get("url", None), url)
        self.assertEqual(result.get("title", None), title)
        clicks = long(result.get("clicks"))

        #test the clicks increases when navigate
        response = self.browse(shorturl)
        self.assertTrue(response.status == 301)
        
        result = client.url_stats(shorturl)
        clicks_new = long(result.get("clicks"))
        
        self.assertTrue(clicks_new == clicks + 1)

if __name__ == '__main__':
    unittest.main()


testing/test_yourlsClient.py
# test_yourlsClient.py
#  - tests for the python yourls client
#
# Copyright 2011, Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Author:
#       Tim Flink <tflink@redhat.com>


import pytest
import yourls.client
from yourls import YourlsError, YourlsOperationError
import json

test_baseurl = 'http://localhost/yourls/'
test_apiurl = test_baseurl + 'yourls-api.php'
test_user = 'username'
test_pass = 'password'
test_urltitle = 'testshort'
test_numclicks = 3

test_url1 = 'http://testhost.fedoraproject.org/dir/subdir/file.html'
test_url2 = 'http://testhost.fedoraproject.org/dir/subdir/anotherfile.html'

url_data = {1:test_url1, 2:test_url2}

#{u'status': u'fail', u'code': u'error:url', u'title': u'Fedora People', u'url': {u'keyword': '', u'title': u'Fedora People', u'url': u'http://tflink.fedorapeople.org/', u'ip': u'192.168.122.1', u'date': u'2011-07-25 13:30:32', u'clicks': u'3'}, u'shorturl': u'http://192.168.122.202/yourls/2', u'message': u'http://tflink.fedorapeople.org/ already exists in database', u'statusCode': 200}
def make_json_response(status, url, shorturl, code = None, message = None,
                        statuscode = None):
    if not message:
        message = 'no message'
    if not statuscode:
        statuscode = 200

    url_data = {'keyword':'', 'title':test_urltitle, 'url':url, 'ip':'127.0.0.1',
                'date':'now', 'clicks':'1'}
    data = {'status':status, 'title':test_urltitle, 'url':url_data,
            'shorturl':shorturl, 'message':message, 'statuscode':statuscode}
    if code is not None:
        data['code'] = code
    return json.dumps(data)

def make_json_error(errorCode, message):
    data = {'message': message, 'errorCode': errorCode}
    return json.dumps(data)

def make_json_expand(baseurl, keyword, longurl):
    data = {'shorturl':(baseurl+str(keyword)), 'longurl':longurl,
            'keyword':keyword, 'message':'success', 'statusCode':200}
    return json.dumps(data)

def make_json_urlstats(baseurl, keyword, longurl, clicks, isError = False):
    if not isError:
        url_data = {'title' : 'Testing Link Title' , 'url' : longurl,
                    'ip' : '127.0.0.1', 'shorturl' : test_baseurl + str(keyword),
                    'clicks' : clicks }
        data = {'message' : 'success', 'link' : url_data, 'statusCode' : 200}
    else:
        data = {'message' : 'Error: short URL not found', 'statusCode' : 404}

    return json.dumps(data)

def mock_request_error(args):
        return make_json_error('Invalid username or password', 403)

def mock_request(args):
    if args['action'] == 'shorturl':
        for short, url in url_data.iteritems():
            if args['url'] == url:
                return make_json_response('success', url, test_baseurl + str(short))
    elif args['action'] == 'expand':
        for short, url in url_data.iteritems():
            if args['shorturl'] == short:
                return make_json_expand(test_baseurl, short, url)
    elif args['action'] == 'url-stats':
        for short, url in url_data.iteritems():
            if args['shorturl'] == short:
                return make_json_urlstats( test_baseurl, short, url, test_numclicks,
                                           isError = False)

def mock_short_keyworderror(args):
    return make_json_response('fail', test_url1, 1, code = 'error:keyword',
            message = 'Short URL 1 already exists in database or is reserved')

def mock_request_404(args):
    code_name = "statusCode"
    if args['action'] == 'expand':
        code_name = "errorCode"
    return json.dumps({ code_name : 404, 'message' : 'Error: short URL not found',
                        'keyword' : args['shorturl']})

class TestYourlsClient():

    def setup_method(self, method):
        self.testclient = yourls.client.YourlsClient(test_apiurl, test_user, test_pass)

    def test_shorten_url(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request)

        ref_shorturl = test_baseurl + '1'
        test_url = 'http://testhost.fedoraproject.org/dir/subdir/file.html'
        shorturl = self.testclient.shorten(test_url)

        assert shorturl == ref_shorturl

    def test_shorten_second_url(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request)

        ref_shorturl = test_baseurl + '2'
        test_url = 'http://testhost.fedoraproject.org/dir/subdir/anotherfile.html'
        shorturl = self.testclient.shorten(test_url)

        assert shorturl == ref_shorturl

    def test_make_shorten_args(self):
        ref_args = {'username':test_user, 'password':test_pass, 'format':'json',
                    'action':'shorturl', 'url':test_url1}
        ref_newargs = {'action':'shorturl', 'url':test_url1}

        test_args = self.testclient._make_args(ref_newargs)

        assert ref_args == test_args

    def test_shorten_used_keyword(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_short_keyworderror)
        with pytest.raises(YourlsOperationError):
            self.testclient.shorten(test_url1)

    def test_shorten_invalid_user(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request_error)
        with pytest.raises(YourlsOperationError):
            self.testclient.shorten(test_url1)

    def test_expand_url(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request)
        ref_shorturl = 1

        test_longurl = self.testclient.expand(ref_shorturl)

        assert test_longurl == url_data[ref_shorturl]

    def test_expand_nonexistant_url(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request_404)
        ref_madeupkeyword= 'blahdontexist'
        with pytest.raises(YourlsOperationError):
            self.testclient.expand(ref_madeupkeyword)

    def test_url_stats(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request)
        ref_shorturl = 1

        test_data = self.testclient.get_url_stats(ref_shorturl)

        assert test_data['clicks'] == test_numclicks

    def test_stats_nonextant_url(self, monkeypatch):
        monkeypatch.setattr(self.testclient, '_send_request', mock_request_404)
        ref_madeupkeyword= 'blahdontexist'
        with pytest.raises(YourlsOperationError):
            self.testclient.get_url_stats(ref_madeupkeyword)

    def test_with_empty_apiurl(self):
        with pytest.raises(YourlsError):
            yourls.client.YourlsClient('', test_user, test_pass)

    def test_with_no_username(self):
        with pytest.raises(YourlsError):
            yourls.client.YourlsClient(test_apiurl, '', test_pass)

    def test_with_no_password(self):
        with pytest.raises(YourlsError):
            yourls.client.YourlsClient(test_apiurl, test_user, '')




searchindex.js
Search.setIndex({objects:{"":{yourls:[0,0,1],client:[0,0,1]},yourls:{YourlsOperationError:[0,2,1],client:[0,0,1],YourlsError:[0,2,1]},"yourls.client.YourlsClient":{get_url_stats:[0,1,1],shorten:[0,1,1],expand:[0,1,1]},"yourls.client":{YourlsClient:[0,3,1]}},terms:{oper:0,all:0,code:0,when:[],documenat:0,paramet:[0,1],password:0,thank:1,easi:0,fix:1,except:[0,1],program:0,wouldn:1,sourc:0,"return":0,string:1,get:0,python:0,initi:1,"import":0,increas:0,yourlsoperationerror:[0,1],now:1,name:1,changelog:[0,1],list:0,token:0,stuff:0,statist:0,small:0,where:1,page:[0,1],pleas:0,creation:1,see:0,download:0,librari:0,out:0,index:[0,1],version:[0,1],"new":0,localhost:0,str:0,fixm:0,base:0,repositori:0,releas:1,org:0,thrown:[0,1],about:0,autoqa:0,yourlserror:0,place:1,isn:0,improv:1,chang:[0,1],com:0,origin:0,modul:[0,1],shorturl:0,api:[0,1],messag:0,yourlsclient:0,given:0,git:0,from:0,creat:0,support:1,yourlskeyworderror:1,been:0,custom:0,interfac:0,handl:1,more:1,tflink:0,option:0,form:0,search:[0,1],"short":0,greatli:1,none:0,keyword:0,provid:0,empti:1,can:0,behavior:1,error:[0,1],tar:0,dure:0,get_url_stat:0,indic:[0,1],tabl:[0,1],tarbal:0,fedorahost:0,check:0,perform:0,titl:[0,1],shorten:0,detail:0,how:0,exampl:0,test:1,instead:0,document:0,simpl:0,conflict:[],complet:0,http:0,usernam:0,clone:0,"class":0,rais:[0,1],php:0,apiurl:0,expand:0,yourl:0,github:0,url:[0,1],request:0,issu:1,davmuz:1,client:0,thi:0,time:1},objtypes:{"0":"py:module","1":"py:method","2":"py:exception","3":"py:class"},titles:["YOURLS Client","Changelog"],objnames:{"0":"Python module","1":"Python method","2":"Python exception","3":"Python class"},filenames:["index","changelog"]})

yourls/__init__.py
# yourls
#  - Python client for yourls
#
# Copyright 2011, Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Author:
#       Tim Flink <tflink@redhat.com>

__version__ = "0.3.0"

class YourlsError(Exception):
    '''Base exception for all Yourls exceptions'''
    def __init__(self, message):
        self.message = message
    def __str__(self):
        return self.message

class YourlsOperationError(YourlsError):
    '''Error during URL operations'''
    def __init__(self, url, message):
        self.url = url
        self.message = message
    def __str__(self):
        return repr('Error with url \'%s\' - %s' % (self.url, self.message))



yourls/client.py
# client.py
#  - Python client for yourls
#
# Copyright 2011, Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Author:
#       Tim Flink <tflink@redhat.com>

"""
.. module:: yourls.client
   :synopsis: A simple client for the YOURLS URL shortener

.. moduleauthor:: Tim Flink <tflink@redhat.com>
"""

import sys
import json
import urllib
import urllib2

from yourls import __version__, YourlsError, YourlsOperationError

LIMIT=100

class YourlsClient():

    def __init__(self, apiurl, username=None, 
                password=None, token=None, format='json'):
        """The use of a username/password combo or a signature token is required

        :param apiurl: The location of the api php file
        :param username: The username to login with (not needed with signature token)
        :param password: The password to login with (not needed with signature token)
        :param token: The signature token to use (not needed with username/password combo)
        :param format: The default format for all messages
        
        :throws: YourlsError for incorrent parameters

        """
        self.data_format = format

        if not apiurl:
            raise YourlsError("An api url is required")
        self.apiurl = apiurl

        if not username or not password:
            if not token:
                raise YourlsError("username and password or signature token are required")
            else:
                self.std_args = {'signature' : token, 'format' : self.data_format}
        else:
            self.username = username
            self.password = password
            self.std_args = {'username':self.username, 'password':self.password,
                             'format':self.data_format}


    def _send_request(self, args):
        """Encapsulates the actual sending of a request to a YOURLS instance

        :param args: The arguments to send to YOURLS

        """
        urlargs = urllib.urlencode(self._make_args(args))
        req = urllib2.Request(self.apiurl)
        req.add_data(urlargs)
        r = urllib2.urlopen(req)
        data = r.read()
        return data


    def _make_args(self, new_args):
        """Convenience method for putting args into the proper format

        :param new_args: Dictionary containing the args to pass on

        """
        return dict(self.std_args.items() + new_args.items())


    def _base_request(self, args, url):
        """Encapsulates common code and error handling for the access methods

        :param args: The arguments to send to YOURLS
        :param url: The url (short or long) arg being used in the request
        :raises: YourlsOperationError

        """
        try:
            data = json.loads(self._send_request(args))
        except urllib2.URLError as error:
            raise YourlsOperationError(url, str(error))

        if 'errorCode' in data:
            raise YourlsOperationError(url, data['message'])

        return data


    def shorturl(self, url, keyword=None, title=None):
        """Request a shortened URL from YOURLS with an optional keyword request

        :param url: The URL to shorten
        :type url: str
        :param keyword: The keyword keyword to request
        :type keyword: str
        :param title: Use the given title instead of download it from the URL, this will increase performances
        :type title: str
        :returns: str -- The short URL
        :raises: YourlsOperationError

        """
        args = {'action':'shorturl', 'url':url}

        if keyword:
            args['keyword'] = keyword

        if title:
            args['title'] = title

        # shorten
        raw_data = self._base_request(args, url)

        # parse result
        if raw_data['status'] == 'fail' and raw_data['code'] == 'error:keyword':
            raise YourlsOperationError(url, raw_data['message'])

        if not 'shorturl' in raw_data:
            raise YourlsOperationError(url, 'Unknown error: %s' % raw_data['message'])

        return raw_data['shorturl']

    shorten = shorturl

    def expand(self, shorturl):
        """Expand a shortened URL to its original form

        :param shorturl: The URL to expand
        :returns: str -- The expanded URL
        :raises: YourlsOperationError

        """
        args = {'action' : 'expand', 'shorturl' : shorturl, 'format' : 'json'}

        raw_data = self._base_request(args, shorturl)

        if not 'longurl' in raw_data:
            raise YourlsOperationError(shorturl, raw_data['message'])

        return raw_data['longurl']


    def url_stats(self, shorturl):
        """Get statistics about a shortened URL

        :param shorturl: The URL to expand
        :returns: a list of stuff - FIXME, this isn't complete
        :raises: YourlsOperationError

        """

        args = {'action' : 'url-stats', 'shorturl' : shorturl, 'format' : 'json'}

        raw_data = self._base_request(args, shorturl)

        if raw_data['statusCode'] != 200:
            raise YourlsOperationError(shorturl, raw_data['message'])

        return raw_data['link']
    get_url_stats = url_stats

    def stats(self, filter, limit=LIMIT):
        """Get statistics about your links

        :param filter: either "top", "bottom", "rand" or "last"
        :type filter: str
        :param limit: maximum number of links to return
        :type limit: int
        :returns: a list of stuff - FIXME, this isn't complete
        :raises: YourlsOperationError

        """

        args = {'action' : 'stats', 'filter' : filter, 
                'limit':limit, 'format' : 'json'}

        raw_data = self._base_request(args, shorturl)

        if raw_data['statusCode'] != 200:
            raise YourlsOperationError(shorturl, raw_data['message'])

        return raw_data["stats"]

    def db_stats(self):
        """Get global link and click count

        :returns: a list of stuff - FIXME, this isn't complete
        :raises: YourlsOperationError

        """

        args = {'action' : 'db-stats', 'format' : 'json'}
        raw_data = self._base_request(args, shorturl)
        return raw_data

def get_server(apiurl, **kwargs):
    kw = dict(apiurl=apiurl)
    for key in ["username", "password", "token"]:
        value = kwargs.get(key, None)
        if value:
            kw[key] = value
    return YourlsClient(**kw)

def shorturl(url, keyword, title=None, server=None, **kwargs):
    if server is None:
        server = get_server(**kwargs)
    return server.shorturl(url, keyword=keyword, title=title)
shorten = shorturl

def expand(shorturl, server=None, **kwargs):
    if server is None:
        server = get_server(**kwargs)
    return server.expand(shorturl)

def url_stats(shorturl, server=None, **kwargs):
    if server is None:
        server = get_server(**kwargs)
    return server.url_stats(shorturl)
get_url_stats = url_stats

def stats(filter="top", limit=LIMIT, server=None, **kwargs):
    if server is None:
        server = get_server(**kwargs)
    return server.stats(filter, limit)

def db_stats(server=None, **kwargs):
    if server is None:
        server = get_server(**kwargs)
    return server.db_stats()

def set_yourls_parser(parser):
    parser.add_argument("-s", "--apiurl", metavar="uri",
        default="http://localhost/yourls/yourls-api.php",
        help="Yourls API URL. default=%(default)s")
    parser.add_argument("-T", "--token", help="Token")
    parser.add_argument("-u", "--username", help="Username")
    parser.add_argument("-p", "--password", help="Password")

def main():
    """Yourls command line access"""
    
    import argparse

    parser = argparse.ArgumentParser(
        description='Yourls Client',
        version="%(prog)s " + __version__
        )

    subparsers = parser.add_subparsers(dest="command")

    #
    #shorturl command
    parser_shorturl = subparsers.add_parser('shorturl', 
        help="get short URL for a link")

    set_yourls_parser(parser_shorturl)
    parser_shorturl.add_argument("-k", "--keyword", type=unicode, 
        help="optional keyword for custom short URLs")
    parser_shorturl.add_argument("-t", "--title", type=unicode, 
        help="title")
    parser_shorturl.add_argument("url", type=unicode, 
        help="the url to shorten")
    parser_shorturl.set_defaults(func=shorturl)

    #expand command
    parser_expand = subparsers.add_parser('expand', help="get long URL of a shorturl")
    set_yourls_parser(parser_expand)
    parser_expand.add_argument("url", type=unicode, 
        help="the shorturl to expand (can be either 'abc' or 'http://site/abc')")
    parser_expand.set_defaults(func=expand)

    #url_stats command
    parser_url_stats = subparsers.add_parser(
        'url-stats', help="get stats about one short URL")
    set_yourls_parser(parser_url_stats)
    parser_url_stats.add_argument("shorturl", type=unicode, 
        help="the shorturl for which to get stats (can be either 'abc' or 'http://site/abc')")
    parser_url_stats.set_defaults(func=url_stats)

    #stats command
    parser_stats = subparsers.add_parser(
        'stats', help="get stats about your links")
    set_yourls_parser(parser_stats)
    parser_stats.add_argument("-l","--limit", type=int, default=LIMIT,
        help="maximum number of links to return. default=%(default)s")
    parser_stats.add_argument("-f", "--filter", type=str, default="top",
        choices=['top', 'bottom', 'rand', 'last'],
        help="the filter: either 'top', 'bottom' , 'rand' or 'last'. default=%(default)s")
    parser_stats.set_defaults(func=stats)
    
    #db_stats command
    parser_db_stats = subparsers.add_parser(
        'db-stats', help="get global link and click count")
    set_yourls_parser(parser_db_stats)
    parser_db_stats.set_defaults(func=db_stats)

    args = parser.parse_args()
    try:
        #fargs, fkwargs = config.func_args(args)
        result = args.func(**vars(args))
    except Exception, exc:
        print >> sys.stderr, "FATAL! %s executing %s" % (
            exc, args.func.func_name)
        return 1
    else:
        return result

if __name__ == '__main__':
    exit(main())

